internal/app/machined/pkg/runtime/v1alpha1/bootloader/sdboot/efivars.go

```diff
@@ -81,5 +74,99 @@ func WriteVariable(c efivario.Context, name, value string) error {
 
 	out = append(out[:n], 0, 0)
 
-	return c.Set(name, SystemdBootGUID, efivario.BootServiceAccess|efivario.RuntimeAccess|efivario.NonVolatile, out)
+	return efivarfs.Write(efivarfs.ScopeSystemd, name, efivarfs.AttrBootserviceAccess|efivarfs.AttrRuntimeAccess|efivarfs.AttrNonVolatile, out)
+}
+
+// CreateBootEntry creates a UEFI boot entry named "Talos Linux UKI" and sets it as the first in the `BootOrder`
+// with the specified install disk and architecture.
+// The entry will point to the SystemdBoot PE binary located at the specified install disk path.
+//
+//nolint:gocyclo
+func CreateBootEntry(installDisk, sdBootFilePath string) error {
+	// mount EFI vars as rw
+	if err := unix.Mount("efivarfs", constants.EFIVarsMountPoint, "efivarfs", unix.MS_REMOUNT, ""); err != nil {
+		return err
+	}
+
+	defer unix.Mount("efivarfs", constants.EFIVarsMountPoint, "efivarfs", unix.MS_REMOUNT|unix.MS_RDONLY, "") //nolint:errcheck
+
+	rawBootOrderData, _, err := efivarfs.Read(efivarfs.ScopeGlobal, "BootOrder")
+	if err != nil {
+		return fmt.Errorf("failed to read BootOrder: %w", err)
+	}
+
+	bootOrder, err := efivarfs.UnmarshalBootOrder(rawBootOrderData)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal BootOrder: %w", err)
+	}
+
+	// Let's assume we are adding a new boot entry for Talos Linux UKI.
+	talosBootIndex := len(bootOrder)
+
+	for _, idx := range bootOrder {
+		bootEntry, err := efivarfs.GetBootEntry(int(idx))
+		if err != nil {
+			return fmt.Errorf("failed to get boot entry %d: %w", idx, err)
+		}
+
+		if bootEntry.Description == TalosBootEntryDescription {
+			// If we already have a Talos Linux UKI boot entry, we will use its index.
+			// This allows us to update the existing entry instead of creating a new one.
+			talosBootIndex = int(idx)
+
+			break
+		}
+	}
+
+	info, err := blkid.ProbePath(installDisk, blkid.WithSkipLocking(true))
+	if err != nil {
+		return fmt.Errorf("failed to probe install disk %q: %w", installDisk, err)
+	}
+
+	efiPartInfo := xslices.Filter(info.Parts, func(part blkid.NestedProbeResult) bool {
+		return part.PartitionLabel != nil && *part.PartitionLabel == constants.EFIPartitionLabel
+	})
+
+	if len(efiPartInfo) == 0 {
+		return fmt.Errorf("EFI partition not found on install disk %q", installDisk)
+	}
+
+	partitionUUID := efiPartInfo[0].PartitionUUID
+
+	if partitionUUID == nil {
+		return fmt.Errorf("EFI partition UUID not found on install disk %q", installDisk)
+	}
+
+	if err := efivarfs.SetBootEntry(talosBootIndex, &efivarfs.LoadOption{
+		Description: TalosBootEntryDescription,
+		FilePath: efivarfs.DevicePath{
+			&efivarfs.HardDrivePath{
+				PartitionNumber:     1,
```

Comment:
Is this dependent on ESP being the first partition? Perhaps it'd be better to get this from somewhere